In [16]:
import pandas as pd
from scipy import  stats
import numpy as np
from scipy.stats import chi2_contingency

In [8]:
df = pd.read_csv("Database_practice\house_price_Dống-Da_Hà-Nội_subdata.csv")
df.head()

area    price  postDate land_certificate house_direction balcony_direction  \
0  50.0  14700.0     44197            So do             NaN               NaN   
1  57.0   4200.0     44265            So do             Nam               Nam   
2  62.0  11000.0     44268            So do             NaN          dong-Nam   
3  26.0   2900.0     44292            So do             NaN               NaN   
4  40.0   3950.0     44292            So do             NaN               NaN   

   toilet  bedroom  floor        lat        long property_type  
0     4.0      6.0    6.0  21.015698  105.824583     trong ngo  
1     3.0      8.0    4.0  21.025208  105.833111     trong ngo  
2     6.0      5.0    6.0  21.025379  105.801011     trong ngo  
3     4.0      3.0    4.0  21.014325  105.834918     trong ngo  
4     NaN      NaN    5.0  21.009862  105.838083     trong ngo

In [9]:
df1 = df.replace([np.inf, -np.inf], np.nan)
df1.head()

area    price  postDate land_certificate house_direction balcony_direction  \
0  50.0  14700.0     44197            So do             NaN               NaN   
1  57.0   4200.0     44265            So do             Nam               Nam   
2  62.0  11000.0     44268            So do             NaN          dong-Nam   
3  26.0   2900.0     44292            So do             NaN               NaN   
4  40.0   3950.0     44292            So do             NaN               NaN   

   toilet  bedroom  floor        lat        long property_type  
0     4.0      6.0    6.0  21.015698  105.824583     trong ngo  
1     3.0      8.0    4.0  21.025208  105.833111     trong ngo  
2     6.0      5.0    6.0  21.025379  105.801011     trong ngo  
3     4.0      3.0    4.0  21.014325  105.834918     trong ngo  
4     NaN      NaN    5.0  21.009862  105.838083     trong ngo

In [10]:
df1.fillna(method="ffill", inplace = True)

# Hãy tiến hành kiểm định tương quan giữa các thuộc tính sau với mức ý nghĩa 5%. Trong đó giá nhà (price) đang được để theo đơn vị (nghìn đồng)


# 1.Giữa giá nhà và diện tích có tương quan với nhau?
 

In [11]:
# Sử dụng kiểm định pearson để kiểm định tương quan giữa hai thuộc tính định lượng

In [13]:
print  ("hệ số tương quan và pvalue tương ứng là: \n",stats.pearsonr(df1.price, df1.area))

hệ số tương quan và pvalue tương ứng là: 
 (0.22970731862850846, 1.7458335920070289e-09)


Tiến hành kiểm định
- Giả thuyết không: Giữa giá nhà và diện tích không tương quan với nhau (H0 ở đây chính là r = 0 (giữa 2 thuộc tính không có mối tương quan)
- Giả thuyết đối: Giữa giá nhà và diện tích có tương quan với nhau
ĐỌC kết quả
- Với alpha = 0.05, ta có pvalue = 1.7458335920070289e-09 < alpha (phủ nhận H0: 2 cái này ko có mỗi tương quan) => 2 thuộc tính giá nhà và diện tích này tương quan với nhau
- Với r = 0.2297 > 0 2 thuộc tính này tương quan với nhau

# 2.Giữa giá nhà và tọa độ địa lý (lat, long) có tương quan với nhau

In [ ]:
# Sử dụng kiểm định pearson để kiểm định tương quan giữa hai thuộc tính định lượng

In [14]:
# Giữa giá nhà và tọa độ địa lý (lat)
print  ("hệ số tương quan và pvalue tương ứng là: \n",stats.pearsonr(df1.price, df1.lat))

hệ số tương quan và pvalue tương ứng là: 
 (-0.05811435352233809, 0.13262275510358396)


- Giữa giá nhà và tọa độ địa lý (lat)
*****Tiến hành kiểm định****
- Giả thuyết không: Giữa giá nhà và lat không tương quan với nhau (H0 ở đây chính là r = 0 (giữa 2 thuộc tính không có mối tương quan)
- Giả thuyết đối: Giữa giá nhà và lat có tương quan với nhau 
****ĐỌC kết quả****
- Với alpha = 0.05, ta có pvalue = 0.132 > alpha (chấp nhận H0: 2 cái này ko có mối tương quan)
- Với r = -0.05 < 0 2 thuộc tính này không tương quan với nhau

In [15]:
# Giữa giá nhà và tọa độ địa lý (long)
print  ("hệ số tương quan và pvalue tương ứng là: \n",stats.pearsonr(df1.price, df1.long))

hệ số tương quan và pvalue tương ứng là: 
 (0.06644958649780451, 0.08543607092904575)


- Giữa giá nhà và tọa độ địa lý (long)
*Tiến hành kiểm định*
- Giả thuyết không: Giữa giá nhà và long không tương quan với nhau (H0 ở đây chính là r = 0 (giữa 2 thuộc tính không có mối tương quan)
- Giả thuyết đối: Giữa giá nhà và long có tương quan với nhau
*ĐỌC kết quả*
- Với alpha = 0.05, ta có pvalue = 0.08 > alpha (chấp nhận H0: 2 cái này ko có mối tương quan)
- Với r = 0.06 > 0 rất ít, 2 thuộc tính này gần như không tương quan với nhau

# 3.Giữa thuộc tính land_certificate và property_type có tương quan với nhau


In [ ]:
# Kiểm định Chi bình phương để kiểm định tương quan giữa hai thuộc tính định tính

In [29]:
data = df1.groupby('property_type').count()
data

area  price  postDate  land_certificate  house_direction  \
property_type                                                             
mat pho         222    222       222               222              222   
trong ngo       449    449       449               449              448   

               balcony_direction  toilet  bedroom  floor  lat  long  
property_type                                                        
mat pho                      222     222      222    222  222   222  
trong ngo                    448     449      449    449  449   449

In [30]:
chi, p, dof, expected = chi2_contingency (data)
print ('Chi-square value = ', chi)
print ('p value = ', p)
print ('Degree of freedom = ', dof)
print ('expected table = ', expected)

Chi-square value =  0.0012077300009639635
p value =  1.0
Degree of freedom =  10
expected table =  [[222.06017075 222.06017075 222.06017075 222.06017075 221.7292316
  221.7292316  222.06017075 222.06017075 222.06017075 222.06017075
  222.06017075]
 [448.93982925 448.93982925 448.93982925 448.93982925 448.2707684
  448.2707684  448.93982925 448.93982925 448.93982925 448.93982925
  448.93982925]]


* KẾT QUẢ
- H0 không có mối tương quan giữa property_type với các thuộc tính trong đó có land_certificate
- H1 có mối tương quan giữa property_type với các thuộc tính trong đó có land_certificate
- Với pvalue = 1.0 > alpha = 0.05, chấp nhận H0, bác bỏ H1, nên không có mối tương quan giữa property_type với các thuộc tính trong đó có land_certificate

# 4.Hãy  mã hóa lại thuộc tính giá nhà theo đơn vị (nghìn đồng/m2) và sắp xếp giá nhà thành 4 mức tương ứng với các khoảng tứ phân vị. Tiến hành kiểm định tương quan giữa biến giá nhà này với các biến tọa độ vị trí (lat, long)   

In [ ]:
# Kiểm định spearman để kiểm định tương quan giữa một thuộc tính thứ bậc và một thuộc tính liên tục

In [51]:
# Tạo cột giá nhà nghìn đồng/m2
df1['nghinm2'] = round(df1['price'] * 1000000 / df1['area'],0)
df1.head()

area    price  postDate land_certificate house_direction balcony_direction  \
0  50.0  14700.0     44197            So do             NaN               NaN   
1  57.0   4200.0     44265            So do             Nam               Nam   
2  62.0  11000.0     44268            So do             Nam          dong-Nam   
3  26.0   2900.0     44292            So do             Nam          dong-Nam   
4  40.0   3950.0     44292            So do             Nam          dong-Nam   

   toilet  bedroom  floor        lat        long property_type     nghìn/m2  \
0     4.0      6.0    6.0  21.015698  105.824583     trong ngo  294000000.0   
1     3.0      8.0    4.0  21.025208  105.833111     trong ngo   73684211.0   
2     6.0      5.0    6.0  21.025379  105.801011     trong ngo  177419355.0   
3     4.0      3.0    4.0  21.014325  105.834918     trong ngo  111538462.0   
4     4.0      3.0    5.0  21.009862  105.838083     trong ngo   98750000.0   

      nghin/m2      nghinm2  
0  294000000.0  294000000.0  
1   73684211.0   73684211.0  
2  177419355.0  177419355.0  
3  111538462.0  111538462.0  
4   98750000.0   98750000.0

In [52]:
df2 = df1[['nghinm2', 'lat', 'long' ]]
df2.head()

nghinm2        lat        long
0  294000000.0  21.015698  105.824583
1   73684211.0  21.025208  105.833111
2  177419355.0  21.025379  105.801011
3  111538462.0  21.014325  105.834918
4   98750000.0  21.009862  105.838083

In [55]:
df3 = df2.replace([np.inf, -np.inf], 0)
df3.head()

nghinm2        lat        long
0  294000000.0  21.015698  105.824583
1   73684211.0  21.025208  105.833111
2  177419355.0  21.025379  105.801011
3  111538462.0  21.014325  105.834918
4   98750000.0  21.009862  105.838083

In [48]:
df3.shape

(671, 3)

In [58]:
# Lấy gía trị nhỏ nhất của nghinm2
nghinm2s = sorted(df3['nghinm2'].unique(), reverse = True)
rank_nghinm2 = {nghinm2s[i]:i+1 for i in range (len(nghinm2s))}
rank_nghinm2

{8333333333.0: 1,
 3300000000.0: 2,
 1117977528.0: 3,
 1111111111.0: 4,
 1076923077.0: 5,
 1060000000.0: 6,
 847457627.0: 7,
 750000000.0: 8,
 685483871.0: 9,
 680000000.0: 10,
 666666667.0: 11,
 625000000.0: 12,
 621818182.0: 13,
 609756098.0: 14,
 607142857.0: 15,
 576744186.0: 16,
 562500000.0: 17,
 558139535.0: 18,
 551111111.0: 19,
 546428571.0: 20,
 544642857.0: 21,
 544000000.0: 22,
 542857143.0: 23,
 531428571.0: 24,
 528169014.0: 25,
 521538462.0: 26,
 514285714.0: 27,
 511904762.0: 28,
 504807692.0: 29,
 503496503.0: 30,
 492957746.0: 31,
 489690722.0: 32,
 489510490.0: 33,
 480000000.0: 34,
 478365385.0: 35,
 475000000.0: 36,
 470588235.0: 37,
 462500000.0: 38,
 461538462.0: 39,
 457831325.0: 40,
 457142857.0: 41,
 450000000.0: 42,
 441379310.0: 43,
 440170940.0: 44,
 440000000.0: 45,
 439393939.0: 46,
 438461538.0: 47,
 433333333.0: 48,
 420560748.0: 49,
 418181818.0: 50,
 417777778.0: 51,
 411904762.0: 52,
 409090909.0: 53,
 408450704.0: 54,
 407894737.0: 55,
 402352941.0:

In [59]:
rank = pd.Series(rank_nghinm2)
rank

8.333333e+09      1
3.300000e+09      2
1.117978e+09      3
1.111111e+09      4
1.076923e+09      5
               ... 
3.333330e+05    476
2.850000e+05    477
1.956520e+05    478
1.225000e+05    479
0.000000e+00    480
Length: 480, dtype: int64

In [60]:
h = df3['nghinm2']
df3['nghinm2_Rank'] = h.map(rank)
df3

nghinm2        lat        long  nghinm2_Rank
0    294000000.0  21.015698  105.824583           113
1     73684211.0  21.025208  105.833111           396
2    177419355.0  21.025379  105.801011           229
3    111538462.0  21.014325  105.834918           328
4     98750000.0  21.009862  105.838083           351
..           ...        ...         ...           ...
666  139880952.0  21.000823  105.829097           281
667  269230769.0  21.011495  105.816666           141
668  291666667.0  21.011495  105.816666           117
669  551111111.0  21.016106  105.832157            19
670  420560748.0  21.009123  105.824428            49

[671 rows x 4 columns]

In [61]:
# Biến nghinm2_Rank với lat
r, pvalue = stats.spearmanr(df3['nghinm2_Rank'], df3['lat'])
print('r = ', r)
print('pvalue = ', pvalue)

r =  0.04928485022695764
pvalue =  0.20229048815895062


Kết luận:
- Do p_value (0.20) > alpha = 0.05 nên H0 được chấp nhận, HA bị loại bỏ, giữa các thuộc tính không có tương quan, r gần bằng 0, mối quan hệ giữa 2 thuộc tính không chặt chẽ

In [62]:
# Biến nghinm2_Rank với long
r, pvalue = stats.spearmanr(df3['nghinm2_Rank'], df3['long'])
print('r = ', r)
print('pvalue = ', pvalue)

r =  0.22049847786772178
pvalue =  7.821453782581783e-09


Kết luận:

- Do p_value (7.821453782581783e-09) < alpha = 0.05 nên H0 bị loại bỏ, HA được chấp nhận, giữa các thuộc tính có tương quan, r > 0, mối quan hệ giữa 2 thuộc tính không chặt chẽ lắm